### PR TITLE
Chainer multi gpu

### PR DESCRIPTION
**Pull request by kan-bayashi**
Multi gpu implementation with chainer

### CODE REVIEW

**Comment by kan-bayashi** [comment]
I change from `ParallelUpdater` to `MultiProcessParallelUpdater`.
It seems to be working well with high GPU utils.
But the implementation is a little bit complicated.
@sw005320 @ShigekiKarita
If you have ideas to clean up codes, please let met know.

**Comment by kan-bayashi** [comment]
@sw005320 @ShigekiKarita  I implemented following functions:
- warn the use of `--gpu` option in `run.sh`, `asr_train.py`, `asr_recog.py`, `lm_train.py`
- if `--gpu` option is used, `--ngpu` value will be automatically updated
- if use gpu, check whether `CUDA_VISIBLE_DEVICES` is set in `asr_train.py`, `asr_recog.py`, `lm_train.py`
- if the number of gpus is not matched with `CUDA_VISIBLE_DEVICES`, show warning message 
- if the number of gpus > 1, `--ngpu` will be automatically 1 in LM training

File: src/bin/asr_train.py [hunk 1]
```diff
@@ -134,6 +136,24 @@ def main():
...
+            args.ngpu = 1
+
+    # check CUDA_VISIBLE_DEVICES
+    if args.ngpu > 0:
+        cvd = os.environ.get("CUDA_VISIBLE_DEVICES")
+        if cvd is None:
+            logging.warn("CUDA_VISIBLE_DEVICES is not set.")
+        elif args.ngpu != len(cvd.split(",")):
+            logging.warn(cvd)
+            logging.warn(len(cvd.split(",")))
```
Comment (ShigekiKarita) [inline src/bin/asr_train.py]: maybe `logging.error` is better?

File: src/asr/asr_chainer.py [hunk 2]
```diff
@@ -215,31 +389,64 @@ def train(args):
...
     # prepare Kaldi reader
     train_reader = lazy_io.read_dict_scp(args.train_feat)
     valid_reader = lazy_io.read_dict_scp(args.valid_feat)
 
+    # set up training iterator and updater
+    if ngpu <= 1:
+        # make minibatch list (variable length)
+        train = make_batchset(train_json, args.batch_size,
+                              args.maxlen_in, args.maxlen_out, args.minibatches)
+        # hack to make batchsze argument as 1
```
Comment (ShigekiKarita) [inline src/asr/asr_chainer.py]: batchsze -> batchsize

File: src/asr/asr_chainer.py [hunk 3]
```diff
@@ -215,31 +389,64 @@ def train(args):
...
+                                            args.maxlen_in, args.maxlen_out, args.minibatches)]
+
+        # each subset must have same length for MultiprocessParallelUpdater
+        maxlen = max([len(train_subset) for train_subset in train_subsets])
+        for train_subset in train_subsets:
+            if maxlen != len(train_subset):
+                for i in six.moves.xrange(maxlen - len(train_subset)):
+                    train_subset += [train_subset[i]]
+
+        # hack to make batchsze argument as 1
```
Comment (ShigekiKarita) [inline src/asr/asr_chainer.py]: ditto

File: src/asr/asr_chainer.py [hunk 4]
```diff
@@ -215,31 +389,64 @@ def train(args):
...
+            # make minibatch list (variable length)
+            train_subsets += [make_batchset(train_json_subset, args.batch_size,
+                                            args.maxlen_in, args.maxlen_out, args.minibatches)]
+
+        # each subset must have same length for MultiprocessParallelUpdater
+        maxlen = max([len(train_subset) for train_subset in train_subsets])
+        for train_subset in train_subsets:
+            if maxlen != len(train_subset):
+                for i in six.moves.xrange(maxlen - len(train_subset)):
+                    train_subset += [train_subset[i]]
```
Comment (ShigekiKarita) [inline src/asr/asr_chainer.py]: Does a shorter minibatch appear more than a longer minibatch? Is it possible to append empty list?

File: egs/csj/asr1/run.sh [hunk 5]
```diff
@@ -195,9 +212,9 @@ mkdir -p ${expdir}
 
 if [ ${stage} -le 4 ]; then
     echo "stage 4: Network Training"
-    ${cuda_cmd} ${expdir}/train.log \
+    ${cuda_cmd} --gpu ${ngpu} ${expdir}/train.log \
```
Comment (sw005320) [inline egs/csj/asr1/run.sh]: Just make sure that do we need some changes in `cmd.sh` to pass `--gpu ${ngpu}`?

File: src/asr/asr_chainer.py [hunk 6]
```diff
@@ -194,12 +357,23 @@ def train(args):
         logging.info('ARGS: ' + key + ': ' + str(vars(args)[key]))
 
     # Set gpu
-    gpu_id = int(args.gpu)
-    logging.info('gpu id: ' + str(gpu_id))
-    if gpu_id >= 0:
+    ngpu = int(args.ngpu)
```
Comment (sw005320) [inline src/asr/asr_chainer.py]: may not have to cast it to `int()`

File: egs/csj/asr1/run.sh (hunk 5 above)
Comment (kan-bayashi) [reply egs/csj/asr1/run.sh]: In the case of slurm.pl, we do not have to modify `cmd.sh`.

File: src/asr/asr_chainer.py (hunk 4 above)
Comment (kan-bayashi) [reply src/asr/asr_chainer.py]: I've tried but it causes error when the empty batch is used.

**Comment by sw005320** [comment]
What is the progress?
CLSP people will have a problem for this, but I think @bobchennan's solution of adding the CLSP-dependent commands to `run.sh` to get the CUDA_VISBLE_DEVICE is a reasonable solution.

**Comment by kan-bayashi** [comment]
I will add CLSP dependent processing about CLSP.

Are warning messages suitable for you?
And If there is no problem, I will update all of the recipe.

**Comment by sw005320** [comment]
The warning part looks fine.
Please go ahead.

**Comment by kan-bayashi** [comment]
@sw005320 Could you test some recipe in CLSP?

**Comment by sw005320** [comment]
I'm testing, but failed due to 
```
    'NCCL is not enabled. MultiprocessParallelUpdater '
Exception: NCCL is not enabled. MultiprocessParallelUpdater requires NCCL.
Please reinstall chainer after you install NCCL.
(see https://github.com/chainer/chainer#installation).
```

**Comment by sw005320** [comment]
It seems that it's working.
We may need to have some documentation about NCCL (I just got it from the nvidia site also @bobchennan told me where it is). Also we need to remove the following lines in `path.sh`
```
if [[ $(hostname -f) == *.clsp.jhu.edu ]] ; then
    export CUDA_VISIBLE_DEVICES=$(/usr/local/bin/free-gpu)
fi
```

**Comment by kan-bayashi** [comment]
@sw005320 Thank you for your testing.
I updated the use of gpu option and added explanation about NCCL in README.

**Comment by sw005320** [comment]
I think this PR becomes a good shape. @bobchennan can you take a look at it? If it is no problem, I'll merge it. Then, I want to ask you to commit your pytorch multiple GPU function.

**Comment by bobchennan** [comment]
Looks fine for me. But to use it on clsp grid I think we need to modify queue.pl to support multi-gpu.

**Comment by sw005320** [comment]
What part do you need to modify? Since I just used a default `queue.pl` and it worked, but it might be just lucky.

**Comment by bobchennan** [comment]
Yeah you are right. I checked queue.pl in kaldi and it should work. No need for modification.

**Comment by sw005320** [comment]
OK. @kan-bayashi, is there any other issues or things you left? If no, I can merge it.

File: src/asr/asr_chainer.py [hunk 7]
```diff
@@ -215,31 +389,64 @@ def train(args):
...
+            # make minibatch list (variable length)
+            train_subsets += [make_batchset(train_json_subset, args.batch_size,
+                                            args.maxlen_in, args.maxlen_out, args.minibatches)]
+
+        # each subset must have same length for MultiprocessParallelUpdater
+        maxlen = max([len(train_subset) for train_subset in train_subsets])
+        for train_subset in train_subsets:
+            if maxlen != len(train_subset):
+                for i in six.moves.xrange(maxlen - len(train_subset)):
+                    train_subset += [train_subset[i]]
```
Comment (kan-bayashi) [inline src/asr/asr_chainer.py]: @sw005320 my last concern is this part. 
If It is no problem, it can be merged.

**Pull request merged**